File: coinDistribution.py
# Time Complexity : O(N)
# Space Complexity : O(H)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: TreeNode) -> int:
        self.res = 0
        
        self.dfs(root)
        
        return self.res
    
    def dfs(self, root):
        #base
        if not root:
            return 0
        #extra shouldnt be 0
        extra = root.val - 1    #extra = 0
        #logic
        extra += self.dfs(root.left)
        extra += self.dfs(root.right)
        self.res += abs(extra)
        return extra

File: coinDistribution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//Time : O(n) where n is the number of nodes in the tree
//Space : O(h) where h is the height of the tree
class Solution {
public:
    int dfs(TreeNode* root, int &moves){
        //base
        if(!root) return 0;
        //logic
        int left = dfs(root->left,moves);
        int right = dfs(root->right,moves);
        moves += (abs(left) + abs(right));
        return root->val + left + right - 1;
        
    }
    int distributeCoins(TreeNode* root, int moves = 0) {
        dfs(root,moves);
        return moves;
    }
};